# WEO3-SVM

## Hema Sai Ari(59528014)

# Lawn Mower ownership prediction using SVM models

We will predict if the person is owner of lawn mower or not.

## 1. Setup

Import modules

In [1]:
import pandas as pd
from sklearn.svm import SVC
from matplotlib import pyplot as plt
import numpy as np
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder
import pickle

np.random.seed(1)

## 2. Load data




In [2]:
# Uncomment the following snippet of code to debug problems with finding the .csv file path
# This snippet of code will exit the program and print the current working directory.
import os
print(os.getcwd())

C:\Users\arihe\DATA SCIENCE PROGRAMMING\Support Vector Machines


In [3]:
df = pd.read_csv('C:/Users/arihe/Box/SEM-2/DATA SCIENCE PRORAMMING/SIR PYTHON FIES/class-3(More on regression and SVM)/RidingMowers.csv') # let's use the same data as we did in the logistic regression example
df.head(3)

Income  Lot_Size Ownership
0    60.0      18.4     Owner
1    85.5      16.8     Owner
2    64.8      21.6     Owner

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Income     24 non-null     float64
 1   Lot_Size   24 non-null     float64
 2   Ownership  24 non-null     object 
dtypes: float64(2), object(1)
memory usage: 704.0+ bytes


In [5]:
df.isna().sum()

Income       0
Lot_Size     0
Ownership    0
dtype: int64

Converting the string into integer
owner is 1
non owner is 0

To convert the categorical variable into a int we can either use replace or we can use lable encoder

Using replace

df.Ownership=df.Ownership.replace('Owner',1,regex=True)

df.Ownership=df.Ownership.replace('Nonowner',0,regex=True)

In [6]:
#using lable encoder
df['Ownership'] = LabelEncoder().fit_transform(df['Ownership'])
df

Income  Lot_Size  Ownership
0     60.0      18.4          1
1     85.5      16.8          1
2     64.8      21.6          1
3     61.5      20.8          1
4     87.0      23.6          1
5    110.1      19.2          1
6    108.0      17.6          1
7     82.8      22.4          1
8     69.0      20.0          1
9     93.0      20.8          1
10    51.0      22.0          1
11    81.0      20.0          1
12    75.0      19.6          0
13    52.8      20.8          0
14    64.8      17.2          0
15    43.2      20.4          0
16    84.0      17.6          0
17    49.2      17.6          0
18    59.4      16.0          0
19    66.0      18.4          0
20    47.4      16.4          0
21    33.0      18.8          0
22    51.0      14.0          0
23    63.0      14.8          0

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Income     24 non-null     float64
 1   Lot_Size   24 non-null     float64
 2   Ownership  24 non-null     int32  
dtypes: float64(2), int32(1)
memory usage: 608.0 bytes


In [8]:
# split the data into validation and training set with an 70-30 ratio
train_df, test_df = train_test_split(df, test_size=0.3)

# to reduce repetition in later code, create variables to represent the columns
# that are our predictors and target
target = 'Ownership'
predictors = list(df.columns)
predictors.remove(target)

In [9]:
X_train = train_df[predictors]
X_test = test_df[predictors]
y_train = train_df[target]
y_test = test_df[target]

## 3. Model the data

First, let's create a dataframe to load the model performance metrics into.

In [10]:
performance = pd.DataFrame({"model": [], "Accuracy": [], "Precision": [], "Recall": [], "F1": []})

### 3.1 Fit a SVM classification model using linear kernal

In [11]:
svm_lin_model = SVC(kernel="linear",probability=True)
svm_lin = svm_lin_model.fit(X_train, np.ravel(y_train))

In [12]:
model_preds = svm_lin_model.predict(X_test)
c_matrix = confusion_matrix(y_test, model_preds)
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"linear svm", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

### 3.2 Fit a SVM classification model using rbf kernal

In [13]:
svm_rbf_model = SVC(kernel="rbf", C=10, gamma='scale',probability=True)
svm_rbf = svm_rbf_model.fit(X_train, np.ravel(y_train))

In [14]:
model_preds = svm_rbf_model.predict(X_test)
c_matrix = confusion_matrix(y_test, model_preds)
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"rbf svm", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

### 3.3 Fit a SVM classification model using polynomial kernal

In [15]:
svm_poly_model = SVC(kernel="poly", degree=3, coef0=1, C=10,probability=True)
svm_poly = svm_poly_model.fit(X_train, np.ravel(y_train))

In [16]:
model_preds = svm_poly_model.predict(X_test)
c_matrix = confusion_matrix(y_test, model_preds)
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"poly svm", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## 4.0 Summary



In [17]:
performance

model  Accuracy  Precision    Recall        F1
0  linear svm     1.000   1.000000  1.000000  1.000000
0     rbf svm     0.750   0.666667  0.666667  0.666667
0    poly svm     0.875   1.000000  0.666667  0.800000

## Conclusion

As we less observation few models over fit the data. Here we can observe that the linear model is overfitting the data.And we dont the other factors like marking cost, profit gained if the coustomer takes a mower, etc. So to decide the best model for the data lets compare the accuracy. 

If we compare the accuracy of all the models we can clearlly observe that linear model is over fitting the data so the next best model is the poly svm with an accuracy of 0.875.

### Lets save our best model 

We can save our model and can use it later

In [18]:

#save model
pickle.dump(svm_rbf, open('we03_svm.pkl', "wb"))
